### PR TITLE
Oauth bridge has authenticate method

### PR DESCRIPTION
**Pull request by wmathurin**
Application no longer does a login by default, instead the javascript needs to call authenticate if it needs to be authenticated.

First I had a shouldAuthenticate method in SalesforceReactActivity, but it was creating problems when the login automatically started in the middle of the react page running. 

This implementation allows for react native driven splash screen (see comment below).

### CODE REVIEW

File: libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceOAuthPlugin.java [hunk 1]
```diff
@@ -105,15 +105,8 @@ protected void authenticate(final CallbackContext callbackContext) throws JSONEx
...
-        Log.i("SalesforceOAuthPlugin.getAuthCredentials", "getAuthCredentials called");
-    	JSONObject credentials = ((SalesforceDroidGapActivity) cordova.getActivity()).getJSONCredentials();
-        if (credentials == null) {
-            Log.w("SalesforceOAuthPlugin.getAuthCredentials", "getAuthCredentials failed - never authenticated");
-            callbackContext.error("Never authenticated");
-        } else {
-            Log.i("SalesforceOAuthPlugin.getAuthCredentials", "getAuthCredentials successful");
-            callbackContext.success(credentials);
-        }    
+        Log.i(TAG, "getAuthCredentials called");
```
Comment (wmathurin) [inline libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceOAuthPlugin.java]: Refactored to keep the cordova plugin and the react bridge as similar as possible.

File: libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/ui/SalesforceDroidGapActivity.java [hunk 2]
```diff
@@ -534,30 +542,6 @@ private void addSidCookieForDomain(CookieManager cookieMgr, String domain, Strin
        cookieMgr.setCookie(domain, cookieStr);
    }
 
-   /**
```
Comment (wmathurin) [inline libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/ui/SalesforceDroidGapActivity.java]: Moved to RestClient - now being called from two places: react and cordova

File: libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/bridge/SalesforceNetReactBridge.java [hunk 3]
```diff
@@ -150,15 +151,7 @@ private RestRequest prepareRestRequest(ReadableMap args) throws UnsupportedEncod
...
-        if (restClient == null) {
-            UserAccount account = SalesforceSDKManager.getInstance().getUserAccountManager().getCurrentUser();
-            if (account == null) {
-                restClient = SalesforceSDKManager.getInstance().getClientManager().peekUnauthenticatedRestClient();
-            } else {
-                restClient = SalesforceSDKManager.getInstance().getClientManager().peekRestClient(account);
-            }
-        }
-        return restClient;
+        return ((SalesforceReactActivity) getCurrentActivity()).getRestClient();
```
Comment (wmathurin) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/bridge/SalesforceNetReactBridge.java]: Using the rest client owned by the react activity

File: libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java [hunk 4]
```diff
@@ -102,50 +96,80 @@ public void onPause() {
...
+                     SalesforceReactActivity.this.client = client;
+
+
+	                /*
+                     * Do a cheap REST call to refresh the access token if needed.
+                     * If the login took place a while back (e.g. the already logged
+                     * in application was restarted), then the returned session ID
+                     * (access token) might be stale.
+                     */
+                     client.sendAsync(RestRequest.getRequestForResources(ApiVersionStrings.getVersionNumber(SalesforceReactActivity.this)), new RestClient.AsyncRequestCallback() {
```
Comment (wmathurin) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: We do the same for cordova.
If you use the session token directly, it better be fresh.
That code could also be moved to the getAuthCredentials method.

File: reactnative/ReactNativeTemplateApp/js/index.android.js [hunk 5]
```diff
@@ -37,22 +37,40 @@ var {
     Navigator
 } = React;
 var forceClient = require('./react.force.net.js');
+var oauth = require('./react.force.oauth.js');
 
 var App = React.createClass({
+    getInitialState: function() {
```
Comment (wmathurin) [inline reactnative/ReactNativeTemplateApp/js/index.android.js]: NEW: Your app needs to call authenticate before doing anything that requires authentication

**Comment by wmathurin** [comment]
One sample app updated here: https://github.com/wmathurin/MyUserPicReactNative/pull/2

File: libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/ui/SalesforceDroidGapActivity.java [hunk 6]
```diff
@@ -312,40 +305,55 @@ public void authenticatedRestClient(RestClient client) {
...
-                        	if (callbackContext != null) {
-                            	callbackContext.error(exception.getMessage());
-                        	}
-                        }
-                    });
-	            }
+						@Override
+						public void onError(Exception exception) {
+							if (callbackContext != null) {
+								callbackContext.error(exception.getMessage());
```
Comment (bhariharan) [inline libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/ui/SalesforceDroidGapActivity.java]: Weird spacing between ```callbackContext``` and ```.error...```.

File: libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/bridge/SalesforceNetReactBridge.java (hunk 3 above)
Comment (bhariharan) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/bridge/SalesforceNetReactBridge.java]: Does this tie the bridge to ```SalesforceReactActivity```, or was that always the case anyway?

File: libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java [hunk 7]
```diff
@@ -26,39 +26,46 @@
...
     protected void onCreate(Bundle savedInstanceState) {
+        Log.i(TAG, "onCreate called");
         super.onCreate(savedInstanceState);
 
+        // Get clientManager
+        clientManager = buildClientManager();
+
         // Gets an instance of the passcode manager.
         passcodeManager = SalesforceSDKManager.getInstance().getPasscodeManager();
-        userSwitchReceiver = new ActivityUserSwitchReceiver();
```
Comment (bhariharan) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: What happened to the user switcher receiver?

Comment (wmathurin) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: There is a story for providing a sf account manager bridge for react. So right now, there is no way of doing it in a react native application. Until then, I don't think we need the user switcher in SalesforceReactActivity.

Comment (bhariharan) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: Gotcha. Maybe we should add a ```TODO:``` somewhere in this class, so we don't forget to add it back.

**Comment by bhariharan** [comment]
Looks good to me otherwise.

File: libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java (hunk 7 above)
Comment (wmathurin) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: Done here https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/1095/commits/9e7d69000cb272c94f883d966054fb8e6453254a

**Pull request merged**